### ISSUE TITLE
[Feature Request] Cache/LoadingCache to support ad-hoc loader function for bulk get

### ISSUE THREAD

**Issue by DitongDing** [google/guava#3866]
Hi Team

TL;DR: Does it make sense to add a "ImmutableMap<K, V> getAll(Iterable<? extends K> keys, LoaderFunction<K, V>)" interface into either Cache or LoadingCache?

I am using the Guava cache in one of my program to implement the "if cached, return; otherwise create, cache and return" pattern.

For single get request, both work perfectly:
* [Cache with ad-hoc Callable](https://github.com/google/guava/blob/master/guava/src/com/google/common/cache/Cache.java#L99)
* [LoadingCache interface with CacheLoader defined at the beginning](https://github.com/google/guava/blob/master/guava/src/com/google/common/cache/LoadingCache.java#L67)

But for bulk get, it became a bit tricky:

* Cache supports ad-hoc loader function for single get only
* LoadingCache supports bulk get, but the CacheLoader cannot be replaced by ad-hoc loader function.

Essentially, what I want to archive is to pass-in some context into the loader to generate cache value. This has pros/cons though:

Pros:
* More flexible to bulk get/create cached value.

Cons:
* Since context can be passed in, it has the risk of "generating different cache value for the same key". It is also possible today, though, by letting CacheLoader to generate random value.
* Possible dead lock if the passed-in function's context has dependency on cache itself. The same problem for  Cache's "V get(K key, Callable<? extends V> loader)".
  * Btw, Cache's interface takes in a supplier (aka Callable), it is already taking context (aka "key") unavoidably...

Will this be a useful feature for more people to consume?

Thanks,
Ditong Ding

**Comment by ben-manes** [comment]
A bulk get in a synchronous cache is tricky to do in a way that blocks cache stampedes. In Guava (and Caffeine) it doesn't, they are documented as performing a `getAllPresent` followed by a `loadAll`, and inserting / replacing the loaded entries. This allows for concurrent usage, but also allows for redundant work.

In Caffeine's `AsyncCache` variants, the entry is a future. It can therefore insert a proxy future into the holes, load the entries, and fill the futures. Any other lookup will retrieve one of these futures, so there is a nice performance win. There are options to use bulk loader methods in that library's interfaces.

Unfortunately doing this in a synchronous cache is tricky. Guava's design would allow it because it's hash table entries are similar to futures (see `LoadingValueReference`). The above optimization was proposed but not accepted, and now wouldn't be implemented due to maintenance concerns. Caffeine is more limited here by building on top of ConcurrentHashMap, where native synchronization is incompatible with locking a dynamic list of objects.

For now I'd recommend using Caffeine's `AsyncCache` as the best compromise to have both an ad hoc loader function that protects against cache stampedes. You can use the `synchronous()` view to simulate a synchronous cache, even though it will diverge in minute ways. Bringing this optimization to the synchronous case hasn't seemed feasible in the design approaches I've brainstormed.

**Comment by DitongDing** [comment]
Thanks for the recommendation! Close the issue :-)
